corporate/lib/stripe.py

```diff
@@ -1666,6 +1669,16 @@ def make_end_of_cycle_updates_if_needed(
         if event_in_next_billing_cycle and last_ledger_entry is not None:
             licenses_at_next_renewal = last_ledger_entry.licenses_at_next_renewal
             assert licenses_at_next_renewal is not None
+
+            if (
+                plan.tier == CustomerPlan.TIER_SELF_HOSTED_LEGACY
+                and plan.end_date == next_billing_cycle
+                and plan.status == CustomerPlan.ACTIVE
+            ):
+                plan.status = CustomerPlan.ENDED
+                plan.save(update_fields=["status"])
+                return None, None
```

Comment:
I am not sure what you have in mind, but set `plan.status` to `CustomerPlan.SWITCH_PLAN_TIER_AT_PLAN_END` in case we want user to switch to a new plan after the current plan ends.

Reply:
Ahh, yeah, that's a good point; given that we signal a next plan through `plan.status`, this does seem safe. OK.

But we we also probably need to call `do_change_plan_type` to move them to `PLAN_TYPE_SELF_MANAGED` from `PLAN_TYPE_SELF_MANAGED_LEGACY` when doing this downgrade....

Actually should this just be calling `self.process_downgrade(plan, background_update=True)`?

One other thought -- I think this logic shouldn't be conditional on `plan.tier`? 

Oh or maybe the bug is that `migrate_customer_to_legacy_plan` incorrectly set the status to `CustomerPlan.ACTIVE` instead of `CustomerPlan.DOWNGRADE_AT_END_OF_CYCLE`? Since I just proposed code that would exactly reimplement that status.

Reply:
Ooops! I agree with, this should just call `self.process_downgrade(plan, background_update=True)`.

Setting `CustomerPlan.DOWNGRADE_AT_END_OF_CYCLE` as status is confusing for someone new looking at the code since we use `end_of_cycle` in reference to `billing_cycle_anchor + billing_frequency` which is not the case here and doesn't give us any advantage here.